[FLINK-4191] Expose shard information in kinesis deserialization schema

rmetzger opened the pull request:
@tzulitai please review the change.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/serialization/KinesisDeserializationSchema.java:
```diff
@@ -43,7 +44,7 @@
 	 * @throws IOException
 	 */
-	T deserialize(byte[] recordKey, byte[] recordValue, String stream, String seqNum, long approxArrivalTimestamp) throws IOException;
+	T deserialize(byte[] recordKey, byte[] recordValue, String stream, Shard shard, String seqNum, long approxArrivalTimestamp) throws IOException;
```
Can you also update the Javadoc for this method?

tzulitai commented:
LGTM, +1 to merge the change after travis turns green + minor Javadoc update.

tzulitai commented:
Btw, should we also be adding the info from `Record#getPartitionKey()` to the deserialization schema too? Might be off the scope of this issue, but might as well now that we're on to this?

rmetzger commented:
I added the javadoc and did some other changes:
- Commented out the `isEndOfStream()` method, because its not respected.
- Re-ordered the arguments of the deser schema. The first group is record specific, the second one stream/shard specific.
I also removed the `Shard` from the list of arguments again. If we are going to shade kinesis at some point, we'll need to break the interface. Also, we are relying on amazon not to change the class. I think now we are on the save side.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:
```diff
@@ -206,7 +206,13 @@ private void deserializeRecordForCollectionAndUpdateState(UserRecord record)
+				record.getSequenceNumber(),
+				approxArrivalTimestamp,
+				subscribedShard.getStreamName(),
+				subscribedShard.getShard().getShardId());
```
Are these supposed to be indented with two tabs?

tzulitai replied:
`subscribedShard.getStreamName()` and `subscribedShard.getShard().getShardId()` will be the same for all records collected in the while loop. Probably won't have a big performance difference, but perhaps it would be better to have `final` values for these two to point out this fact?

rmetzger replied:
Fixed the indention.
I don't think that having a separate final variable for this will improve the performance a lot. The methods are computing anything.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/serialization/KinesisDeserializationSchema.java:
```diff
@@ -37,13 +37,15 @@
 	 * @param seqNum the sequence number of this record in the Kinesis shard
 	 * @param approxArrivalTimestamp the server-side timestamp of when Kinesis received and stored the record
+	 * @param stream the name of the Kinesis stream that this record was sent to
+	 * @param shardId The identifier of the shard the record was sent to.
```
nit: some param descriptions end with period, some don't. I think the others all don't end with period, let's keep consistent with that.

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/serialization/KinesisDeserializationSchemaWrapper.java:
```diff
@@ -37,15 +37,15 @@ public KinesisDeserializationSchemaWrapper(DeserializationSchema<T> deserializat
 	public boolean isEndOfStream(T nextElement) {
 		return deserializationSchema.isEndOfStream(nextElement);
-	}
+	} */
```
Not sure if we should have commented out code. Open up a JIRA for this instead?

tzulitai commented:
Good points on not using Amazon's `Shard` and `Record`, did not think of that!

tzulitai commented on flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/serialization/KinesisDeserializationSchema.java:
```diff
@@ -52,5 +54,5 @@
 	 * @return true if the element signals end of stream, false otherwise
 	 */
-	boolean isEndOfStream(T nextElement);
+	// TODO ADD SUPPORT FOR boolean isEndOfStream(T nextElement);
```
Not sure if we should have commented out code. Open up a JIRA for this instead?

tzulitai replied:
Ah, I just noticed the JIRA for this. Let's leave it as it is then, and uncomment it for the fix on that JIRA.

rmetzger commented:
I fixed the remaining comments. I'll merge the changes tomorrow.

tzulitai commented:
Just realized that the `byte[] recordKey` and `String partitionKey` are actually the same only in different form, so now we have duplicate info in the deser schema. I think we can keep the `String` one, since it's the type we get from AWS API for the key. Opening a hotfix for this.
